### ISSUE TITLE
cabal uses too much memory when solving for setup dependencies

### ISSUE THREAD

**Issue by grayjay** [haskell/cabal#2899]
Some combinations of setup and library dependencies cause the solver to use more and more memory.  I ran `cabal install --dry-run --max-backjumps -1` with this cabal file for five minutes with about 16GB of memory.  I used GHC 7.10.2.

```
name:                example
version:             0.1.0.0
cabal-version:       >=1.8
build-type:          Custom

custom-setup
  setup-depends:       aeson, hackage-server, yi

library
  build-depends:       aeson, mtl, leksah
```

![cabal-memory-usage](https://cloud.githubusercontent.com/assets/4276753/10780987/483a8bba-7d00-11e5-8734-47b18ef62a04.png)

My best guess is that it is related to the `addLinking` function creating linked nodes from existing unlinked nodes.  When a node is referenced by a later sibling, it probably isn't garbage collected immediately after it is traversed. https://github.com/haskell/cabal/blob/c5b90c4d157f64156b6c8efde2aafb1ac86a851b/cabal-install/Distribution/Client/Dependency/Modular/Linking.hs#L67

**Comment by grayjay** [comment]
I encountered this issue again while testing with a long-running command: `cabal install leksah yesod hackage-server`

This uses GHC 8.0.1.
![issue-2899](https://cloud.githubusercontent.com/assets/4276753/16710313/c72c463c-45de-11e6-88f8-a76b40c555f5.png)

The -v3 log shows that the solver chose to link a setup dependency near the top of the tree (level 8).  It didn't return to that level until around the time that the memory usage dropped.  When it descended again, it used a different goal order and chose the setup dependencies much farther down in the tree.  I think that the spike in memory usage can be explained by the solver retaining the whole subtree starting where it chose the first setup dependency.

**Comment by grayjay** [comment]
Fixed in #4110.
